packages/SwingSet/docs/vat-environment.md

```diff
@@ -0,0 +1,113 @@
+
+To avoid confusion, the shim use a regular expression to reject code that looks like it is performing a direct eval. This regexp is not complete (it is possible to trick it into performing a direct eval anyway), but that's ok, because our goal is merely to guide people away from confusing behaviors early in their development process.
+
+This regexp will falsely reject occurrences inside static strings and comments.
+
+### Frozen globalThis
+
+Vats run in a Compartment whose `globalThis` object is frozen. A mutable global object would provide an "ambient communication channel": one side sets `globalThis.heyBuddyAreYouOutThere = 'exfiltrated message'` and the other side periodically reads it. This would violate one of the rules of object-capability security: objects may only communicate through references.
+
+Vats can create new Compartments, and they get to decide whether the new `globalThis` is frozen or not. If the Compartment is meant to support ocap security, the creator should do `harden(c.globalThis)` before loading any code into it.
+
+### Frozen Primordials
+
+SES freezes the "primordials": the built-in JavaScript objects like `Object`, `Array`, and `RegExp`, as well as their prototype chains. This prevents malicious code from changing the behavior of built-ins in surprising ways (imagine `Array.prototype.push` being changed to deliver a copy of its argument to the attacker, or simply ignore certain values). It also prevents the use of e.g. `Object.heyBuddy` as an ambient communication channel.
+
+Both frozen primordials and a frozen `globalThis` will break a few JS libraries that add new features to built-in objects (shims/polyfills). For shims which merely add properties to `globalThis`, it may be possible to load these in a new non-frozen Compartment. Shims which modify primordials can only work if you build new (mutable) wrappers around the default primordials and allow the shims to modify those wrappers instead.
+
+## Library Compatibility
+
+Vat code can use `import` or `require()` to import other libraries, as long as those libraries consist of purely JS code, and are compatible with the SES environment. This includes a significant portion of the NPM registry.
+
+However, many NPM packages use built-in Node.js modules. If they use these at import time (in their top-level code), the package cannot be used by vat code: the vat will fail to load at all. If they wait until runtime to use the built-in features, then the package can be loaded, but might fail later when a function is invoked that accesses the missing functionality. As a result, some NPM packages may be partially compatible: you can use them as long as you don't invoke certain features.
```

Comment:
"can be loaded" -> "can load"

"partially compatible:" Change colon to semi-colon.